src/components/ReportsPage/HeatmapPage/HeatmapUtils.ts

```diff
@@ -94,25 +116,128 @@ const groupReportDataBy = <T extends GroupByValues>(
+    case GroupByOption.ENTITY:
       return _groupBy(scores, (s: ScorecardServiceScore) => s.serviceId);
     case GroupByOption.SERVICE_GROUP:
       return groupReportDataBy(scores, 'tags');
     case GroupByOption.TEAM:
       return groupReportDataBy(scores, 'teams');
     case GroupByOption.LEVEL:
       return groupReportDataBy(scores, 'ladderLevels');
+    case GroupByOption.DOMAIN:
+      return groupReportDataBy(scores, 'domains', domainTagByEntityId);
     default:
       return {};
   }
 };
 
+
+interface HierarchyNode {
+  node: {
+    tag: string;
+  }
+  orderedChildren: HierarchyNode[];
+}
+
+export const hierarchyNodeFlatChildren = (node: HierarchyNode): string[] => {
+  return node.orderedChildren.flatMap((child) => [child.node.tag, ...hierarchyNodeFlatChildren(child)]);
+}
+
+export const groupScoresByHierarchies = (groupedScores: StringIndexable<ScorecardServiceScore[]>, nodes: HierarchyNode[]) => {
+  const hierarchyGroupedData = {} as Record<string, ScorecardServiceScore[]>;
+
+  nodes.forEach((parent) => {
+    hierarchyGroupedData[parent.node.tag] = groupedScores[parent.node.tag] ?? [];
+
+    hierarchyNodeFlatChildren(parent).forEach((childTag) => {
+      (groupedScores[childTag] ?? []).forEach((score) => {
+        if (!hierarchyGroupedData[parent.node.tag].find((existingScore) => existingScore.componentRef === score.componentRef)) {
+          hierarchyGroupedData[parent.node.tag].push(score);
+        }
+      });
+    });
+  });
+
+  return hierarchyGroupedData;
+}
+
+export const catalogToRelationsByEntityId = (entitiesByTag: StringIndexable<HomepageEntity>) => {
+  const ownerEmailByEntityId = {} as Record<string, string[]>;
+  const groupTagByEntityId = {} as Record<string, string[]>;
+
+  Object.values(entitiesByTag).forEach((entity) => {
+    if (entity.serviceOwnerEmails.length) {
+      const ownerEmails = entity.serviceOwnerEmails.map(({ email }) => email);
+      if (ownerEmailByEntityId[entity.id]) {
+        ownerEmailByEntityId[entity.id] = uniq([...ownerEmails, ...ownerEmailByEntityId[entity.id]]);
+      } else {
+        ownerEmailByEntityId[entity.id] = ownerEmails;
+      }
+    }
+
+    if (entity.serviceGroupTags) {
+      if (groupTagByEntityId[entity.id]) {
+        groupTagByEntityId[entity.id] = uniq([...entity.serviceGroupTags, ...groupTagByEntityId[entity.id]]);
+      } else {
+        groupTagByEntityId[entity.id] = entity.serviceGroupTags;
+      }
+    }
+  });
+
+  return {
+    ownerEmailByEntityId,
+    groupTagByEntityId
+  };
+}
+
+export const applyScoreFilters = (
+  scores: ScorecardServiceScore[],
+  scoreFilters: ScoreFilters,
+  ownerEmailByEntityId: StringIndexable<string[]>,
+  groupTagByEntityId: StringIndexable<string[]>,
+  domainIdByEntityId: Record<number, number[]>,
+) => {
+  let resultScores = scores;
+
+  if (scoreFilters.serviceIds.length) {
+    resultScores = resultScores.filter((score) => scoreFilters.serviceIds.includes(score.serviceId));
```

Comment:
We're looping over the scores up to 6 times here instead of once. If this endpoint isn't paginated (which I don't think it is), we should definitely optimize.